### PR TITLE
Add module that uses snakeyaml-engine, which only handles yaml 1.2

### PR DESCRIPTION
**Pull request by sideeffffect**
closes https://github.com/circe/circe-yaml/pull/300
closes https://github.com/circe/circe-yaml/pull/301
closes https://github.com/circe/circe-yaml/pull/302

/cc @oyvindberg

### CODE REVIEW

**Comment by oyvindberg** [comment]
Note that i made that commit two minutes after I got it to compile. I gave also never used circe-yaml or snakeyaml before. As such I'd expect it will take some work to finish it. Tests are not green.

Also this seems to be breaking public api, and a breaking change to yaml format. This should at the minimum get a new major version, perhaps it should be a separate artifact, or even a separate library.

**Comment by sideeffffect** [comment]
Maybe we could find a way to do this that _does_ not break the public API.
The `circe-yaml` library would then still be not well configurable, but maybe it's better to have API continuity than configurability :+1:

**Comment by oyvindberg** [comment]
I'd say the opposite, let's make it very obvious and clear. This update can and will interpret arbitrary yaml differently

Also the new snakeyaml library is very configurable with config objects. Let's rather use those, so for instance the printer can become much simpler. 

The only thing we really need to provide is a Node <-> JSON bijection and some syntax

**Comment by sideeffffect** [comment]
Related issue on BitBucket: snakeyaml-engine Cannot parse integers encoded in hexadecimal
https://bitbucket.org/snakeyaml/snakeyaml-engine/issues/38/cannot-parse-integers-encoded-in

**Review by jeffmay — APPROVED** [review]

**Comment by sideeffffect** [comment]
Hi @jeffmay , would you like to merge this yourself of can I do it?
Just make sure, are you aware that this is a breaking change for the library, right?

File: src/main/scala/io/circe/yaml/Parser.scala
Comment (LeifW) [inline src/main/scala/io/circe/yaml/Parser.scala:66]: If all this does is call the superclass method, it doesn't seem like this override is necessary?

Comment (sideeffffect) [reply src/main/scala/io/circe/yaml/Parser.scala:66]: Good point. Removed :+1:

File: build.sbt
Comment (LeifW) [inline build.sbt:21]: 2.5 is out

File: src/main/scala/io/circe/yaml/Parser.scala
Comment (LeifW) [inline src/main/scala/io/circe/yaml/Parser.scala:70]: Does this thread-safety comment still apply

File: src/main/scala/io/circe/yaml/Parser.scala
Comment (LeifW) [inline src/main/scala/io/circe/yaml/Parser.scala:158]: This builder uses a fluent pattern (each of these "set..." calls returns `this`) - meaning there's no need for this block with `val builder = ...` and a bunch of side-effecting statments. This could be written e.g.
```scala
extends Parser(
  LoadSettings.builder
    .setFoo(...)
    .setBar(...)
    .build
  )
  ```

File: src/main/scala/io/circe/yaml/Parser.scala
Comment (LeifW) [inline src/main/scala/io/circe/yaml/Parser.scala:71]: See the `if node.getValue.startsWith("0x") || node.getValue.contains("_")` check below - snakeyaml-engine doesn't seem handle hexadecimal-formatted numbers as far as I can tell. Those just get tagged / handled as string.

File: src/main/scala/io/circe/yaml/Parser.scala
Comment (LeifW) [inline src/main/scala/io/circe/yaml/Parser.scala:37]:
```scala
  new Compose(settings).composeReader(reader)
```
There's a utl method that defines all this + the `.getSingleNode` call in the form of [Compose.composeReader](
https://javadoc.io/static/org.snakeyaml/snakeyaml-engine/2.5/org/snakeyaml/engine/v2/api/lowlevel/Compose.html#composeReader-java.io.Reader-)

File: src/main/scala/io/circe/yaml/Printer.scala
Comment (LeifW) [inline src/main/scala/io/circe/yaml/Printer.scala:121]: As this class does nothing but call the class it subclasses, it seems unnecessary.

File: build.sbt
Comment (sideeffffect) [reply build.sbt:21]: Updated, thanks

File: src/main/scala/io/circe/yaml/Parser.scala
Comment (sideeffffect) [reply src/main/scala/io/circe/yaml/Parser.scala:70]: I don't know, to be honest. And since I don't know, I'm picking the conservative option of keeping things as they were.

File: src/main/scala/io/circe/yaml/Parser.scala
Comment (sideeffffect) [reply src/main/scala/io/circe/yaml/Parser.scala:158]: Good point, I've simplified the code, thanks.

File: src/main/scala/io/circe/yaml/Parser.scala
Comment (sideeffffect) [reply src/main/scala/io/circe/yaml/Parser.scala:71]: Not now, but maybe in the future
https://bitbucket.org/snakeyaml/snakeyaml-engine/issues/38/cannot-parse-integers-encoded-in

File: src/main/scala/io/circe/yaml/Printer.scala
Comment (sideeffffect) [reply src/main/scala/io/circe/yaml/Printer.scala:121]: Good point, removed :+1:

**Comment by sideeffffect** [comment]
I wasn't able to implement all the suggestions from the comments, so I've at least explained the reason in the comments.

Otherwise this is ready to be merged @jeffmay

**Comment by sideeffffect** [comment]
I'll rework this into 2 separate modules, one for 1.1, the other for 1.2.

**Comment by sideeffffect** [comment]
@jeffmay This is ready. Can you please merge this and release a new version? (I made sure that this is 100 % backwards compatible.)

File: circe-yaml-v12/src/main/scala/io/circe/yaml/v12/ParserImpl.scala
Comment (LeifW) [inline circe-yaml-v12/src/main/scala/io/circe/yaml/v12/ParserImpl.scala:74]: As the snakeyaml-engine readme says, "Only [JSON Schema](http://yaml.org/spec/1.2/spec.html#id2803231) is supported." - that is, it doesn't support any literal formats beyond what's already in JSON. You can inspect the snakeyaml-engine impl of `construct` here and see that it's not doing anything fancier than e.g. `String.toInt`. So we can remove this entire `case Tag.INT if node.getValue ... =>` block - it's not doing anything other than wasting CPU cycles and confusing / misleading future maintainers.
https://bitbucket.org/snakeyaml/snakeyaml-engine/src/master/#markdown-header-snakeyaml-engine-features

File: circe-yaml-v12/src/main/scala/io/circe/yaml/v12/ParserImpl.scala
Comment (LeifW) [inline circe-yaml-v12/src/main/scala/io/circe/yaml/v12/ParserImpl.scala:87]: See above on how snakeyaml-engine "only supports JSON schema". We have no need for calling`construct` method on literals, as snakeyaml-engine doesn't support any more literal formats than JSON itself already does.

File: circe-yaml-v12/src/main/scala/io/circe/yaml/v12/ParserImpl.scala
Comment (sideeffffect) [reply circe-yaml-v12/src/main/scala/io/circe/yaml/v12/ParserImpl.scala:74]: Do you think this code would be useful once we add Core Schema support to snakeyaml-engine?
https://bitbucket.org/snakeyaml/snakeyaml-engine/issues/38/cannot-parse-integers-encoded-in

**Comment by sideeffffect** [comment]
I'll go ahead and merge this before any more merge conflicts appear. We can keep polishing this after it is in master

**Pull request merged**
